From 6aef92f89723a798ecc2be281b90f0066572d48d Mon Sep 17 00:00:00 2001
From: "debing.sun" <debing.sun@redis.com>
Date: Wed, 13 Aug 2025 16:09:44 +0800
Subject: [PATCH] Fix MurmurHash64A overflow in HyperLogLog with 2GB+ entries

The MurmurHash64A function in hyperloglog.c used an int parameter for length,
causing integer overflow when processing PFADD entries larger than 2GB.
This could lead to server crashes.

Changed the len parameter from int to size_t to properly handle
large inputs up to SIZE_MAX in HyperLogLog operations.
Refer to the implementation in facebook/mcrouter@2dbee3d/mcrouter/lib/fbi/hash.c#L54

(cherry picked from commit 681246c2488ee1ab6d7e667512d878c35b612fca)
---
 src/hyperloglog.c          | 2 +-
 tests/support/util.tcl     | 2 +-
 tests/unit/hyperloglog.tcl | 8 ++++++++
 3 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/hyperloglog.c b/src/hyperloglog.c
index 53276d129..83ead1f7a 100644
--- a/src/hyperloglog.c
+++ b/src/hyperloglog.c
@@ -394,7 +394,7 @@ static char *invalid_hll_err = "-INVALIDOBJ Corrupted HLL object detected";
  * It was modified for Redis in order to provide the same result in
  * big and little endian archs (endian neutral). */
 REDIS_NO_SANITIZE("alignment")
-uint64_t MurmurHash64A (const void * key, int len, unsigned int seed) {
+uint64_t MurmurHash64A (const void * key, size_t len, unsigned int seed) {
     const uint64_t m = 0xc6a4a7935bd1e995;
     const int r = 47;
     uint64_t h = seed ^ (len * m);
diff --git a/tests/support/util.tcl b/tests/support/util.tcl
index 891ce6730..4199b858f 100644
--- a/tests/support/util.tcl
+++ b/tests/support/util.tcl
@@ -987,7 +987,7 @@ proc get_nonloopback_client {} {
 }
 
 # The following functions and variables are used only when running large-memory
-# tests. We avoid defining them when not running large-memory tests because the 
+# tests. We avoid defining them when not running large-memory tests because the
 # global variables takes up lots of memory.
 proc init_large_mem_vars {} {
     if {![info exists ::str500]} {
diff --git a/tests/unit/hyperloglog.tcl b/tests/unit/hyperloglog.tcl
index 24bac26e5..8be591210 100644
--- a/tests/unit/hyperloglog.tcl
+++ b/tests/unit/hyperloglog.tcl
@@ -262,4 +262,12 @@ start_server {tags {"hll"}} {
         r pfadd hll 1 2 3
         assert {[r getrange hll 15 15] eq "\x80"}
     }
+
+    test {PFADD with 2GB entry should not crash server due to overflow in MurmurHash64A} {
+        r config set proto-max-bulk-len 3221225472
+        r config set client-query-buffer-limit 3221225472
+        r write "*3\r\n\$5\r\nPFADD\r\n\$3\r\nhll\r\n"
+        write_big_bulk 2147483648;
+        r ping
+    } {PONG} {large-memory}
 }
-- 
2.25.1

